add node provisioning probe

Omarabdul3ziz opened the pull request:
# in progress
to add :
- [ ] handle dedicated/rented nodes
- [ ] deploy/record in batches
- [ ] handle zoslight nodes
- [ ] after each cycle & on startup: delete all contracts - in case something stuck
- [ ] add timescale to the live `processor_db` instead
- [ ] use better queries, `sqlc`?
- [ ] add auto retention based on configured window
- [ ] add more rates to score: healthy? deployment duration?

Omarabdul3ziz committed:
Add initial implementation of provision-probe service

Omarabdul3ziz committed:
support concurrent deployments with semaphore

Omarabdul3ziz committed:
improve the deployment logic

Omarabdul3ziz committed:
Add API server with scoring endpoints and update configuration structure

- Introduced a new API server with endpoints for retrieving top scores and individual node scores.
- Updated configuration structure to include API settings and probe parameters.
- Added new dependencies for routing and logging.
- Refactored database interaction to support scoring logic.
- Enhanced example configuration file to reflect new settings.

Omarabdul3ziz committed:
Enhance configuration and deployment logic with retry mechanism and graceful shutdown

Omarabdul3ziz committed:
add Swagger documentation, and enhance API responses

Omarabdul3ziz committed:
Add 'provision-probe/' from commit 'e13a3f710f4e706d400bf73274934ec80b870e78'

git-subtree-dir: provision-probe
git-subtree-mainline: ff1e5bfa77b4d0902de5bca243a0ed1b4d8ad379
git-subtree-split: e13a3f710f4e706d400bf73274934ec80b870e78

xmonader commented:
before proceeding please do the following

1- change the name to nodes-deployment-checker
2- switch to gin
3- use swagger/gin integration
4- also stick to gorm, no need to introduce something new in this project

Omarabdul3ziz committed:
rename Deployment Checker service

Omarabdul3ziz committed:
Refactor database interactions to use GORM and implement batch insertion for deployment attempts

- Replaced pgx with GORM for database operations.
- Added a new `BatchSize` configuration parameter for controlling batch insert size.
- Implemented batch insertion of deployment attempts to improve performance.
- Created a new GORM model for `DeploymentAttempt` to facilitate database operations.
- Updated relevant methods to utilize GORM's features for querying and inserting data.

Omarabdul3ziz committed:
Refactor API to use Gin framework and enhance request handling

Omarabdul3ziz committed:
Add support for deploying light VMs on zoslight nodes &  skip rented and dedicated farm nodes.

Omarabdul3ziz committed:
Add jitter configuration and cleanup on startup for deployment checker

Omarabdul3ziz committed:
Add scoring system for deployment checker

Omarabdul3ziz committed:
Add database retention configuration and implement TimescaleDB retention policies

Omarabdul3ziz committed:
refactor: rearrage the app files

Omarabdul3ziz committed:
chore: refactor and test

Omarabdul3ziz committed:
Merge branch 'development' of github.com:threefoldtech/tfgrid-sdk-go into development_provision_probe
